<a href="https://colab.research.google.com/github/Abhhiiissshhek/100_days_of_ML_challenge/blob/main/day-03-feature-engineering/day03_feature_engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **100 days ml challenge**

**Day 03**

In [1]:
import pandas

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pandas.read_csv(url)
print(df.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


**"STEP 1: FIND MISSING VALUES 🔍"**

In [2]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**177 | 687 | 2 MISSING**

Now let us calulate the damage % (percentage)

In [3]:
(df.isnull().sum() / len(df)) * 100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

**THE DROP DECISION (Cabin)**

">50% MISSING = DROP it 🗑️"

In [4]:
# Cabin is 77% missing - basically useless
df.drop('Cabin', axis=1, inplace=True)
print("Cabin column dropped ☠️")

Cabin column dropped ☠️


we dropped the cabin column

# **THE FILL DECISION (Age & Embarked)**

In [5]:
# Age is 20% missing - fill with MEDIAN
df['Age'].fillna(df['Age'].median(), inplace=True)

# Embarked is 0.2% missing - fill with MODE
df['Embarked'].fillna(df['Embarked'].mode()[0], inplace=True)

/tmp/ipython-input-1486647455.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(df['Age'].median(), inplace=True)
/tmp/ipython-input-1486647455.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try 

### **FILL: MEDIAN for numbers, MODE for categories ✅**

## THE SATISFYING FINISH (Cleaned Data)

In [21]:
df.isnull().sum().sum()

np.int64(0)

**"0 MISSING VALUES 🎉"**

Our data is clean - 0 missing values. But raw columns aren't enough.

# **FEATURE 1: FAMILY SIZE**

In [12]:
# Create Family Size from SibSp + Parch
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1
print(df.head())


   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  FamilySize  
0      0         A/5 21171   7.2500        S           2  
1      0          PC 17599  71.2833        C           2  
2      0  STON/O2. 3101282   7.9250        S           1  
3      0            113803  53.1000        S           2  
4      0            37345

*  Passenger 1: FamilySize = 2  (traveling with spouse)
*  Passenger 6: FamilySize = 1  (alone)



**ALONE vs WITH FAMILY = different survival rates!**

# **Feature 2 : TITLE EXTRACTION**

In [15]:
# Extract Title from Name
df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)


   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  FamilySize Title  
0      0         A/5 21171   7.2500        S           2    Mr  
1      0          PC 17599  71.2833        C           2   Mrs  
2      0  STON/O2. 3101282   7.9250        S           1  Miss  
3      0            113803  53.1000        S           2   M

<>:2: SyntaxWarning: invalid escape sequence '\.'
<>:2: SyntaxWarning: invalid escape sequence '\.'
/tmp/ipython-input-3803275269.py:2: SyntaxWarning: invalid escape sequence '\.'
  df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)



1. Mr     =  517
2. Miss    = 182
3. Mrs     = 125
4. Master  =  40
5. Rare    =  27



TITLE = SOCIAL STATUS = KEY FEATURE!

# **FEATURE 3: AGE GROUPS**

In [17]:
# Create Age Groups
bins = [0, 12, 18, 35, 60, 100]
labels = ['Child', 'Teen', 'Adult', 'Middle-Aged', 'Senior']
df['AgeGroup'] = pandas.cut(df['Age'], bins=bins, labels=labels)

**BINNING continuous data = better patterns 📊**

# **FEATURE 4: IS_ALONE**

In [18]:
# Create IsAlone flag
df['IsAlone'] = (df['FamilySize'] == 1).astype(int)

* IsAlone = 1 → Passenger alone
* IsAlone = 0 → With family

**SIMPLE FEATURES = POWERFUL INSIGHTS!**

## **BEFORE VS AFTER**

In [20]:
print(df.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  FamilySize Title     AgeGroup  \
0      0         A/5 21171   7.2500        S           2    Mr        Adult   
1      0          PC 17599  71.2833        C           2   Mrs  Middle-Aged   
2      0  STON/O2. 3101282   7.9250        S           1  Miss        Adult   
3   

* BEFORE: Age, SibSp, Parch, Name
* AFTER:  AgeGroup, FamilySize, IsAlone, Title

# **4 NEW FEATURES CREATED! 🎉**